src/com/cft/contactmerge/tests/CompareContactPartsTest.java

```diff
@@ -31,43 +30,648 @@ void doNamesMatch_Yes_IgnoreSpaces() {
     }
 
     @Test
-    void doNamesMatch_Yes_IgnoreSpacesAndCaseInsensitive() {
+    void doNamesMatch_Yes_IgnoreSpaces_CaseInsensitive() {
         assertEquals(AnswerType.yes, CompareContactParts.doNamesMatch(" JoHn", "DOE", "JOHN ", " Doe  "));
     }
 
+    //------------------------------------------------------------------------- Punctuation First Names -------------------------------------------------------------------------
```

Comment:
General comments...  
1. You are testing multiple combinations of the same aspect. You shouldn't need to do this and it probably indicates that you need to create separate sub-routines that you can test separately. So... If you had to create your own case insensitive compare, create that method and test that method. You should only need a few tests to verify that it works and then the only thing you need to check is that it is used.

2. In the CaseInsensitve case, I assume the built in Java version does what we need and so we don't even have to test that routine. We only have to verify it is being used property (i.e. two tests... One to make sure it is used on first name comparison, a second to make sure it is used on last names).

This is the same for punctuation, space removal, comparing name parts, etc. If you have to create a method to do those, test those methods individually. Then add a test to make sure that we are using the method on First and Last name comparisons.

Another way to think about it... Test maintenance takes time. Just like our code, we want to write the least amount of code we can to meet the requirements. Each test should cover code that isn't covered by any other test. The only exception is some high level tests that cover specific use cases that our users care about or tests created for bugs that "escaped" our tests in the past. For example, general rules may guarantee that "Lisa Murray-Bilodeau" matches "Lisa Murray" but since matching married and maiden names is of specific interest to ITK, we would want to make sure it shows up in the test list.

Reply:
If we want to split these tests up into sub classes/methods, we should have a discussion on this. I do agree I probably went a bit too far in between the range-edge testing, I was just following patterns in my head to try to cover all the what-if scenarios I could think of. I also believe if we have a stronger pool of testing conditions, it will help me design my code better when I go to refactor it. There are other things I want to keep in mind... such as the cost of writing in two many variables to match the values when the function gets called, along with heavy pattern recognition.

That being said, if I design the code in a way to illuminate the need for testing case+spaces+ect.... separately, I would be more than happy to shave off these tests. I just feel more comfortable if we do have them right now during the startup phase, to ensure this backbone will be as bug-free as possible.

We'll talk more about this in our scrum tomorrow.